_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,303 @@
+You can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:
+
+```Dockerfile
+# Multi-stage build requires Docker 17.05 or higher
+# --------------- Builder stage ---------------
+FROM centos:7 AS builder
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"
+
+RUN yum -y install git java-1.8.0-openjdk-devel
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+RUN cd cruise-control \
+    && ./gradlew jar :cruise-control-metric-reporter:jar
+
+# --------------- Final stage ---------------
+FROM strimzi/kafka:0.13.0-kafka-2.3.0
+COPY --from=builder cruise-control/cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
+```
+Then building and pushing the image:
+```bash
+# When building, tag the image with the name of a container registry
+# that is accessible by the Kubernetes cluster.
+docker build . -t <registry>/<kafka-image-name>
+
+# Push the image to that container registry
+docker push <registry>/<kafka-image-name>
+```
+For the Cruise Control metric reporter to be activated at runtime, its class name must be listed in the 'metrics.reporters' field of the Kafka config.
+For a Strimzi Kafka cluster, you can do this by creating a Kafka custom resource which references the new Kafka 'image' built above and the 'com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter' class name in the 'metrics.reporters' field using the following 'kafka.yaml' file:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    image: <registry>/<kafka-image-name>
+    version: 2.3.0
+    replicas: 3
+    listeners:
+      plain: {}
+      tls: {}
+    config:
+      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
+      offsets.topic.replication.factor: 1
+      transaction.state.log.replication.factor: 1
+      transaction.state.log.min.isr: 1
+      log.message.format.version: "2.3"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+```
+Then applying the Kafka custom resource to the cluster:
+
+```bash
+kubectl apply -f kafka.yaml
+```
+The 'metrics.reporters' field can take a list of comma-separated metric reporter class names.
+At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.
+
+At user-defined time windows, Cruise Control will read these reported metrics from the designated Kafka topic and use them for cluster monitoring and balancing.
+How to configure the connection between Cruise Control and Kafka will be described later in the post.
+
+# Accessing Zookeeper
+
+Cruise Control relies on Zookeeper for the current state and configuration of the Kafka cluster.
```

Comment:
```suggestion
Cruise Control gets the current state and configuration of the Kafka cluster from Zookeeper.
```

I think we should mention that they are in the process of migrating to using the Kafka `AdminClient`.